plutus-metatheory/src/Untyped/RenamingSubstitution.lagda.md

```diff
@@ -28,15 +29,21 @@ lift : {X Y : Set} → Ren X Y → Ren (Maybe X) (Maybe Y)
 lift ρ nothing = nothing
 lift ρ (just x) = just (ρ x)
 
+renList  : {X Y : Set} → Ren X Y → List (X ⊢) → List (Y ⊢)
 ren : {X Y : Set} → Ren X Y → X ⊢ → Y ⊢
-ren ρ (` x)       = ` (ρ x)
-ren ρ (ƛ t)       = ƛ (ren (lift ρ) t)
-ren ρ (t · u)     = ren ρ t · ren ρ u
-ren ρ (force t)   = force (ren ρ t)
-ren ρ (delay t)   = delay (ren ρ t)
-ren ρ (con tcn)   = con tcn
-ren ρ (builtin b) = builtin b
-ren ρ error       = error
+ren ρ (` x)         = ` (ρ x)
+ren ρ (ƛ t)         = ƛ (ren (lift ρ) t)
+ren ρ (t · u)       = ren ρ t · ren ρ u
+ren ρ (force t)     = force (ren ρ t)
+ren ρ (delay t)     = delay (ren ρ t)
+ren ρ (con tcn)     = con tcn
+ren ρ (builtin b)   = builtin b
+ren ρ error         = error
+ren ρ (constr i xs) = constr i (renList ρ xs)
+ren ρ (case x ts)   = case (ren ρ x) (renList ρ ts)
+
+renList ρ [] = []
+renList ρ (x ∷ xs) = ren ρ x ∷ renList ρ xs
```

Comment:
Can one not just do `fmap (ren rho)`?

Reply:
I wish I could, my job would be so much easier. However, the termination checker doesn't like it. AFAIK, the only way around the termination checker is to write a mutually recursive specialised map function, like it's done here.